texinfo/info/info.1
.TH info 1 "7th December 1990"
.SH NAME
info \- GNU's hypertext system
.SH SYNOPSIS
.B info
[
.B \-\-option-name option-value
]
.B \menu-item...
.SH COPYRIGHT
.if n Copyright (C) 1989, 1993 Free Software Foundation, Inc.
.if t Copyright \(co 1989, 1993 Free Software Foundation, Inc.
.SH DESCRIPTION
.LP
The GNU project has a hypertext system called
.I Info
which allows the same source file to be either printed as a
paper manual, or viewed using
.B info.
It is possible to use the
.B info
program from inside Emacs, or to use the stand-alone version described here.
This manual page gives a brief summary of its capabilities.

.SH OPTIONS
.TP
.B \-\-directory directory-path
Add 
.B directory-path 
to the list of directory paths searched when
.B info
needs to find a file.  You may issue 
.B \-\-directory 
multiple times.
Alternatively, you may specify a value for the environment variable
.B INFOPATH;
if 
.B \-\-directory
is not given, the value of 
.B INFOPATH
is used.  The value of 
.B INFOPATH 
is a colon separated list of directory names.  If you do not supply either
.B INFOPATH
or
.B \-\-directory-path,
.B info
uses a default path.
.TP
.B \-f filename
Specify a particular 
.B info
file to visit.  By default,
.B info
visits
the file 
.B dir;
if you use this option, 
.B info
will start with
.B (FILENAME)Top 
as the first file and node.
.TP
.B \-n nodename
Specify a particular node to visit in the initial file that 
.B info
loads.  This is especially useful in conjunction with 
.B \-\-file.
You may specify 
.B \-\-node
multiple times.
.TP
.B -o file
Direct output to
.B file
instead of starting an interactive
.B info
session.
.TP
.B \-h
Produce a relatively brief description of the available 
.B info
options.
.TP
.B \-\-version
Print the version information of 
.B info
and exit.
.TP
.B menu-item
.B info
treats its remaining arguments as the names of menu items.
The first argument is a menu item in the initial node visited,
while the second argument is a menu item in the first argument's
node.  You can easily move to the node of your choice by
specifying the menu names which describe the path to that node.
For example,

.B       info emacs buffers

first selects the menu item 
.B emacs 
in the node 
.B (dir)Top,
and then selects the menu item 
.B buffers
in the node 
.B (emacs)Top.
.SH COMMANDS
When in
.B info
the following commands are available:
.TP
.B h
Invoke the Info tutorial.
.TP
.B ?
Get a short summary of
.B info
commands.
.TP
.B h
Select the
.B info
node from the main directory; this is much more complete than just
using 
.B ?.
.TP
.B Ctrl-g
Abort whatever you are doing.
.TP
.B Ctrl-l
Redraw the screen.
.PP
Selecting other nodes:
.TP
.B n
Move to the "next" node of this node.
.TP
.B p
Move to the "previous" node of this node.
.TP
.B u
Move to this node's "up" node.
.TP
.B m
Pick a menu item specified by name. Picking a menu item causes another
node to be selected. You do not need to type a complete nodename; if
you type a few letters and then a space or tab
.B info
will will try to fill in the rest of the nodename. If you ask for further
completion without typing any more characters you'll be given a list
of possibilities; you can also get the list with 
.B ?. 
If you type a few characters and then hit return
.B info
will try to do a completion, and if it is ambigous use the first possibility.
.TP
.B f
Follow a cross reference. You are asked for the name of the reference,
using command completion as for
.B m.
.TP
.B l
Move to the last node you were at.
.PP
Moving within a node:
.TP
.B Space
Scroll forward a page.
.TP
.B DEL
Scroll backward a page.
.TP
.B b
Go to the beginning of this node.
.PP
Advanced commands:
.TP
.B q
Quit 
.B info.
.TP
.B 1
Pick first item in node's menu.
.TP
.B 2 \-\- 5
Pick second ... fifth item in node's menu.
.TP
.B g
Move to node specified by name.	You may include a filename as well,
as 
.B (FILENAME)NODENAME.
.TP
.B s
Search through this 
.B info
file for a specified string, and select the node in which 
the next occurrence is found.
.TP
.B M-x print-node
Pipe the contents of the current node through the command in the
environment variable 
.B INFO_PRINT_COMMAND.
If the variable does not exist, the node is simply piped to 
.B lpr.
.SH ENVIRONMENT
.TP
.B INFOPATH
A colon-separated list of directories to search for 
.B info
files.  Used if 
.B \-\-directory
is not given.
.TP
.B INFO_PRINT_COMMAND
The command used for printing.
.SH SEE ALSO
.BR emacs (1)
.SH AUTHOR
.RS
Brian Fox, Free Software Foundation
.br
bfox@ai.mit.edu
.SH MANUAL AUTHOR
.RS
Robert Lupton; updated by Robert J. Chassell.
.br
rhl@astro.princeton.edu; bob@gnu.ai.mit.edu


gcc/NEWS
Noteworthy changes in GCC version 2.8.0
---------------------------------------

A major change in this release is the addition of a framework for
exception handling, currently used by C++.  Many internal changes and
optimization improvements have been made.  These increase the
maintainability and portability of GCC.  GCC now uses autoconf to
compute many host parameters.

The following lists changes that add new features or targets.

See cp/NEWS for new features of C++ in this release.

New tools and features:

    The Dwarf 2 debugging information format is supported on ELF systems, and
    is the default for -ggdb where available.  It can also be used for C++.
    The Dwarf version 1 debugging format is also permitted for C++, but
    does not work well.

    gcov.c is provided for test coverage analysis and branch profiling
    analysis is also supported; see -fprofile-arcs, -ftest-coverage,
    and -fbranch-probabilities.

    Support for the Checker memory checking tool.

    New switch, -fstack-check, to check for stack overflow on systems that
    don't have such built into their ABI.

    New switches, -Wundef and -Wno-undef to warn if an undefined identifier
    is evaluated in an #if directive.

    Options -Wall and -Wimplicit now cause GCC to warn about implicit int
    in declarations (e.g. `register i;'), since the C Standard committee
    has decided to disallow this in the next revision of the standard;
    -Wimplicit-function-declarations and -Wimplicit-int are subsets of
    this.

    Option -Wsign-compare causes GCC to warn about comparison of signed and
    unsigned values.

    Add -dI option of cccp for cxref.

New features in configuration, installation and specs file handling:

    New option --enable-c-cpplib to configure script.

    Allow --with-cpu on configure command to specify a default CPU.

    The -specs=file switch allows you to override default specs used in
    invoking programs like cc1, as, etc.

    Allow including one specs file from another and renaming a specs
    variable.

    You can now relocate all GCC files with a single environment variable
    or a registry entry under Windows 95 and Windows NT.

Changes in Objective-C:

    The Objective-C Runtime Library has been made thread-safe.

    The Objective-C Runtime Library contains an interface for creating
    mutexes, condition mutexes, and threads; it requires a back-end
    implementation for the specific platform and/or thread package.
    Currently supported are DEC/OSF1, IRIX, Mach, OS/2, POSIX, PCThreads,
    Solaris, and Windows32.  The --enable-threads parameter can be used
    when configuring GCC to enable and select a thread back-end.

    Objective-C is now configured as separate front-end language to GCC,
    making it more convenient to conditionally build it.

    The internal structures of the Objective-C Runtime Library have
    changed sufficiently to warrant a new version number; now version 8.
    Programs compiled with an older version must be recompiled.

    The Objective-C Runtime Library can be built as a DLL on Windows 95
    and Windows NT systems.
    
    The Objective-C Runtime Library implements +load.

The following new targets are supported (see also list under each
individual CPU below):

    Embedded target m32r-elf.
    Embedded Hitachi Super-H using ELF.
    RTEMS real-time system on various CPU targets.
    ARC processor.
    NEC V850 processor.
    Matsushita MN10200 processor.
    Matsushita MN10300 processor.
    Sparc and PowerPC running on VxWorks.
    Support both glibc versions 1 and 2 on Linux-based GNU systems.

New features for DEC Alpha systems:

    Allow detailed specification of IEEE fp support:
      -mieee, -mieee-with-inexact, and -mieee-conformant
      -mfp-trap-mode=xxx, -mfp-round-mode=xxx, -mtrap-precision=xxx
    -mcpu=xxx for CPU selection
    Support scheduling parameters for EV5.
    Add support for BWX, CIX, and MAX instruction set extensions.
    Support Linux-based GNU systems.
    Support VMS.

Additional supported processors and systems for MIPS targets:

    MIPS4 instruction set.
    R4100, R4300 and R5000 processors.
    N32 and N64 ABI.
    IRIX 6.2.
    SNI SINIX.
    
New features for Intel x86 family:

    Add scheduling parameters for Pentium and Pentium Pro.
    Support stabs on Solaris-x86.
    Intel x86 processors running the SCO OpenServer 5 family.
    Intel x86 processors running DG/UX.
    Intel x86 using Cygwin32 or Mingw32 on Windows 95 and Windows NT.

New features for Motorola 68k family:

    Support for 68060 processor.
    More consistent switches to specify processor.
    Motorola 68k family running AUX.
    68040 running pSOS, ELF object files, DBX debugging.
    Coldfire variant of Motorola m68k family.

New features for the HP PA RISC:

    -mspace and m-no-space
    -mlong-load-store and -mno-long-load-store
    -mbig-switch -mno-big-switch

    GCC on the PA requires either gas-2.7 or the HP assembler; for best
    results using GAS is highly recommended.  GAS is required for -g and
    exception handling support.

New features for SPARC-based systems:

    The ultrasparc cpu.
    The sparclet cpu, supporting only a.out file format.

texinfo/info/info.c
/* info.c -- Display nodes of Info files in multiple windows.
   $Id: info.c,v 1.2 1998/03/24 18:06:30 law Exp $

   Copyright (C) 1993, 96, 97 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

   Written by Brian Fox (bfox@ai.mit.edu). */

#include "info.h"
#include "indices.h"
#include "dribble.h"
#include "getopt.h"
#if defined (HANDLE_MAN_PAGES)
#  include "man.h"
#endif /* HANDLE_MAN_PAGES */

/* The version numbers of this version of Info. */
int info_major_version = 2;
int info_minor_version = 17;
int info_patch_level = 0;

/* Non-zero means search all indices for APROPOS_SEARCH_STRING. */
static int apropos_p = 0;

/* Variable containing the string to search for when apropos_p is non-zero. */
static char *apropos_search_string = (char *)NULL;

/* Non-zero means search all indices for INDEX_SEARCH_STRING.  Unlike
   apropos, this puts the user at the node, running info. */
static int index_search_p = 0;

/* Variable containing the string to search for when index_search_p is
   non-zero. */ 
static char *index_search_string = (char *)NULL;

/* Non-zero means print version info only. */
static int print_version_p = 0;

/* Non-zero means print a short description of the options. */
static int print_help_p = 0;

/* Array of the names of nodes that the user specified with "--node" on the
   command line. */
static char **user_nodenames = (char **)NULL;
static int user_nodenames_index = 0;
static int user_nodenames_slots = 0;

/* String specifying the first file to load.  This string can only be set
   by the user specifying "--file" on the command line. */
static char *user_filename = (char *)NULL;

/* String specifying the name of the file to dump nodes to.  This value is
   filled if the user speficies "--output" on the command line. */
static char *user_output_filename = (char *)NULL;

/* Non-zero indicates that when "--output" is specified, all of the menu
   items of the specified nodes (and their subnodes as well) should be
   dumped in the order encountered.  This basically can print a book. */
int dump_subnodes = 0;

/* Structure describing the options that Info accepts.  We pass this structure
   to getopt_long ().  If you add or otherwise change this structure, you must
   also change the string which follows it. */
#define APROPOS_OPTION 1
#define DRIBBLE_OPTION 2
#define RESTORE_OPTION 3
#define IDXSRCH_OPTION 4
static struct option long_options[] = {
  { "apropos", 1, 0, APROPOS_OPTION },
  { "directory", 1, 0, 'd' },
  { "node", 1, 0, 'n' },
  { "file", 1, 0, 'f' },
  { "subnodes", 0, &dump_subnodes, 1 },
  { "output", 1, 0, 'o' },
  { "help", 0, &print_help_p, 1 },
  { "version", 0, &print_version_p, 1 },
  { "dribble", 1, 0, DRIBBLE_OPTION },
  { "restore", 1, 0, RESTORE_OPTION },
  { "index-search", 1, 0, IDXSRCH_OPTION },
  {NULL, 0, NULL, 0}
};

/* String describing the shorthand versions of the long options found above. */
static char *short_options = "d:n:f:o:s";

/* When non-zero, the Info window system has been initialized. */
int info_windows_initialized_p = 0;

/* Some "forward" declarations. */
static void usage (), info_short_help (), remember_info_program_name ();


/* **************************************************************** */
/*                                                                  */
/*                Main Entry Point to the Info Program              */
/*                                                                  */
/* **************************************************************** */

int
main (argc, argv)
     int argc;
     char **argv;
{
  int getopt_long_index;        /* Index returned by getopt_long (). */
  NODE *initial_node;           /* First node loaded by Info. */

  remember_info_program_name (argv[0]);

#ifdef HAVE_SETLOCALE
  /* Set locale via LC_ALL.  */
  setlocale (LC_ALL, "");
#endif

  /* Set the text message domain.  */
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  while (1)
    {
      int option_character;

      option_character = getopt_long
        (argc, argv, short_options, long_options, &getopt_long_index);

      /* getopt_long () returns EOF when there are no more long options. */
      if (option_character == EOF)
        break;

      /* If this is a long option, then get the short version of it. */
      if (option_character == 0 && long_options[getopt_long_index].flag == 0)
        option_character = long_options[getopt_long_index].val;

      /* Case on the option that we have received. */
      switch (option_character)
        {
        case 0:
          break;

          /* User wants to add a directory. */
        case 'd':
          info_add_path (optarg, INFOPATH_PREPEND);
          break;

          /* User is specifying a particular node. */
        case 'n':
          add_pointer_to_array (optarg, user_nodenames_index, user_nodenames,
                                user_nodenames_slots, 10, char *);
          break;

          /* User is specifying a particular Info file. */
        case 'f':
          if (user_filename)
            free (user_filename);

          user_filename = xstrdup (optarg);
          break;

          /* User is specifying the name of a file to output to. */
        case 'o':
          if (user_output_filename)
            free (user_output_filename);
          user_output_filename = xstrdup (optarg);
          break;

          /* User is specifying that she wishes to dump the subnodes of
             the node that she is dumping. */
        case 's':
          dump_subnodes = 1;
          break;

          /* User has specified a string to search all indices for. */
        case APROPOS_OPTION:
          apropos_p = 1;
          maybe_free (apropos_search_string);
          apropos_search_string = xstrdup (optarg);
          break;

          /* User has specified a dribble file to receive keystrokes. */
        case DRIBBLE_OPTION:
          close_dribble_file ();
          open_dribble_file (optarg);
          break;

          /* User has specified an alternate input stream. */
        case RESTORE_OPTION:
          info_set_input_from_file (optarg);
          break;

          /* User has specified a string to search all indices for. */
        case IDXSRCH_OPTION:
          index_search_p = 1;
          maybe_free (index_search_string);
          index_search_string = xstrdup (optarg);
          break;

        default:
          usage ();
        }
    }

  /* If the output device is not a terminal, and no output filename has been
     specified, make user_output_filename be "-", so that the info is written
     to stdout, and turn on the dumping of subnodes. */
  if ((!isatty (fileno (stdout))) && (user_output_filename == (char *)NULL))
    {
      user_output_filename = xstrdup ("-");
      dump_subnodes = 1;
    }

  /* If the user specified --version, then show the version and exit. */
  if (print_version_p)
    {
      printf ("info (GNU %s %s) %s\n", PACKAGE, VERSION, version_string ());
      puts ("Copyright (C) 1997 Free Software Foundation, Inc.\n\
There is NO warranty.  You may redistribute this software\n\
under the terms of the GNU General Public License.\n\
For more information about these matters, see the files named COPYING.");
      exit (0);
    }

  /* If the `--help' option was present, show the help and exit. */
  if (print_help_p)
    {
      info_short_help ();
      exit (0);
    }
  
  /* If the user hasn't specified a path for Info files, default it.  */
  if (!infopath)
    {
      char *path_from_env = getenv ("INFOPATH");

      if (path_from_env)
        {
          unsigned len = strlen (path_from_env);
          /* Trailing : on INFOPATH means insert the default path.  */
          if (len && path_from_env[len - 1] == ':')
            {
              path_from_env[len - 1] = 0;
              info_add_path (DEFAULT_INFOPATH, INFOPATH_PREPEND);
            }
          info_add_path (path_from_env, INFOPATH_PREPEND);
        }
      else
        info_add_path (DEFAULT_INFOPATH, INFOPATH_PREPEND);
    }

  /* If the user specified a particular filename, add the path of that
     file to the contents of INFOPATH. */
  if (user_filename)
    {
      char *directory_name = xstrdup (user_filename);
      char *temp = filename_non_directory (directory_name);

      if (temp != directory_name)
        {
          *temp = 0;
          info_add_path (directory_name, INFOPATH_PREPEND);
        }

      free (directory_name);
    }

  /* If the user wants to search every known index for a given string,
     do that now, and report the results. */
  if (apropos_p)
    {
      info_apropos (apropos_search_string);
      exit (0);
    }

  /* Get the initial Info node.  It is either "(dir)Top", or what the user
     specifed with values in user_filename and user_nodenames. */
  initial_node = info_get_node (user_filename,
                                user_nodenames ? user_nodenames[0] : NULL);

  /* If we couldn't get the initial node, this user is in trouble. */
  if (!initial_node)
    {
      if (info_recent_file_error)
        info_error (info_recent_file_error);
      else
        info_error
          (CANT_FIND_NODE, user_nodenames ? user_nodenames[0] : "Top");
      exit (1);
    }

  /* Special cases for when the user specifies multiple nodes.  If we
     are dumping to an output file, dump all of the nodes specified.
     Otherwise, attempt to create enough windows to handle the nodes
     that this user wants displayed. */
  if (user_nodenames_index > 1)
    {
      free (initial_node);

      if (user_output_filename)
        dump_nodes_to_file
          (user_filename, user_nodenames, user_output_filename, dump_subnodes);
      else
        begin_multiple_window_info_session (user_filename, user_nodenames);

      exit (0);
    }

  /* If the user specified `--index-search=STRING', start the info
     session in the node corresponding to the first match. */
  if (index_search_p)
    {
      int status = 0;

      initialize_info_session (initial_node, 0);

      if (index_entry_exists (windows, index_search_string))
        {
          terminal_clear_screen ();
          terminal_prep_terminal ();
          display_update_display (windows);
          info_last_executed_command = (VFunction *)NULL;

          do_info_index_search (windows, 0, index_search_string);

          info_read_and_dispatch ();

          terminal_unprep_terminal ();

          /* On program exit, leave the cursor at the bottom of the
             window, and restore the terminal IO. */
          terminal_goto_xy (0, screenheight - 1);
          terminal_clear_to_eol ();
          fflush (stdout);
        }
      else
        {
          fputs (_("no entries found\n"), stderr);
          status = 2;
        }

      close_dribble_file (); 
      exit (status);
    }

  /* If there are arguments remaining, they are the names of menu items
     in sequential info files starting from the first one loaded.  That
     file name is either "dir", or the contents of user_filename if one
     was specified. */
  while (optind != argc)
    {
      REFERENCE **menu;
      REFERENCE *entry;
      NODE *node;
      char *arg;
      static char *first_arg = (char *)NULL;

      /* Remember the name of the menu entry we want. */
      arg = argv[optind++];

      if (!first_arg)
        first_arg = arg;

      /* Build and return a list of the menu items in this node. */
      menu = info_menu_of_node (initial_node);

      /* If there wasn't a menu item in this node, stop here, but let
         the user continue to use Info.  Perhaps they wanted this node
         and didn't realize it. */
      if (!menu)
        {
#if defined (HANDLE_MAN_PAGES)
          if (first_arg == arg)
            {
              node = make_manpage_node (first_arg);
              if (node)
                goto maybe_got_node;
            }
#endif /* HANDLE_MAN_PAGES */
          begin_info_session_with_error
            (initial_node, _("There is no menu in this node."));
          exit (0);
        }

      /* Find the specified menu item. */
      entry = info_get_labeled_reference (arg, menu);

      /* If the item wasn't found, search the list sloppily.  Perhaps this
         user typed "buffer" when they really meant "Buffers". */
      if (!entry)
        {
          register int i;
          int best_guess = -1;

          for (i = 0; (entry = menu[i]); i++)
            {
              if (strcasecmp (entry->label, arg) == 0)
                break;
              else
                if (strncasecmp (entry->label, arg, strlen (arg)) == 0)
                  best_guess = i;
            }

          if (!entry && best_guess != -1)
            entry = menu[best_guess];
        }

      /* If we failed to find the reference, start Info with the current
         node anyway.  It is probably a misspelling. */
      if (!entry)
        {
          char *error_message = _("There is no menu item \"%s\" in this node.");

#if defined (HANDLE_MAN_PAGES)
          if (first_arg == arg)
            {
              node = make_manpage_node (first_arg);
              if (node)
                goto maybe_got_node;
            }
#endif /* HANDLE_MAN_PAGES */

          info_free_references (menu);

          /* If we were supposed to dump this node, complain. */
          if (user_output_filename)
            info_error (error_message, arg);
          else
            begin_info_session_with_error (initial_node, error_message, arg);

          exit (0);
        }

      /* We have found the reference that the user specified.  Clean it
         up a little bit. */
      if (!entry->filename)
        {
          if (initial_node->parent)
            entry->filename = xstrdup (initial_node->parent);
          else
            entry->filename = xstrdup (initial_node->filename);
        }

      /* Find this node.  If we can find it, then turn the initial_node
         into this one.  If we cannot find it, try using the label of the
         entry as a file (i.e., "(LABEL)Top").  Otherwise the Info file is
         malformed in some way, and we will just use the current value of
         initial node. */
      node = info_get_node (entry->filename, entry->nodename);

#if defined (HANDLE_MAN_PAGES)
          if ((first_arg == arg) && !node)
            {
              node = make_manpage_node (first_arg);
              if (node)
                goto maybe_got_node;
            }
#endif /* HANDLE_MAN_PAGES */

      if (!node && entry->nodename &&
          (strcmp (entry->label, entry->nodename) == 0))
        node = info_get_node (entry->label, "Top");

    maybe_got_node:
      if (node)
        {
          free (initial_node);
          initial_node = node;
          info_free_references (menu);
        }
      else
        {
          char *temp = xstrdup (entry->label);
          char *error_message;

          error_message = _("Unable to find the node referenced by \"%s\".");

          info_free_references (menu);

          /* If we were trying to dump the node, then give up.  Otherwise,
             start the session with an error message. */
          if (user_output_filename)
            info_error (error_message, temp);
          else
            begin_info_session_with_error (initial_node, error_message, temp);

          exit (0);
        }
    }

  /* If the user specified that this node should be output, then do that
     now.  Otherwise, start the Info session with this node. */
  if (user_output_filename)
    dump_node_to_file (initial_node, user_output_filename, dump_subnodes);
  else
    begin_info_session (initial_node);

  exit (0);
}

/* Return a string describing the current version of Info. */
char *
version_string ()
{
  static char *vstring = (char *)NULL;

  if (!vstring)
    {
      vstring = (char *)xmalloc (50);
      sprintf (vstring, "%d.%d", info_major_version, info_minor_version);
      if (info_patch_level)
        sprintf (vstring + strlen (vstring), "-p%d", info_patch_level);
    }
  return (vstring);
}

/* **************************************************************** */
/*                                                                  */
/*                 Error Handling for Info                          */
/*                                                                  */
/* **************************************************************** */

static char *program_name = (char *)NULL;

static void
remember_info_program_name (fullpath)
     char *fullpath;
{
  char *filename;

  filename = filename_non_directory (fullpath);
  program_name = xstrdup (filename);
}

/* Non-zero if an error has been signalled. */
int info_error_was_printed = 0;

/* Non-zero means ring terminal bell on errors. */
int info_error_rings_bell_p = 1;

/* Print FORMAT with ARG1 and ARG2.  If the window system was initialized,
   then the message is printed in the echo area.  Otherwise, a message is
   output to stderr. */
void
info_error (format, arg1, arg2)
     char *format;
     void *arg1, *arg2;
{
  info_error_was_printed = 1;

  if (!info_windows_initialized_p || display_inhibited)
    {
      fprintf (stderr, "%s: ", program_name);
      fprintf (stderr, format, arg1, arg2);
      fprintf (stderr, "\n");
      fflush (stderr);
    }
  else
    {
      if (!echo_area_is_active)
        {
          if (info_error_rings_bell_p)
            terminal_ring_bell ();
          window_message_in_echo_area (format, arg1, arg2);
        }
      else
        {
          NODE *temp;

          temp = build_message_node (format, arg1, arg2);
          if (info_error_rings_bell_p)
            terminal_ring_bell ();
          inform_in_echo_area (temp->contents);
          free (temp->contents);
          free (temp);
        }
    }
}

/* Produce a very brief descripton of the available options and exit with
   an error. */
static void
usage ()
{
  fprintf (stderr,"%s\n%s\n%s\n%s\n%s\n",
_("Usage: info [-d dir-path] [-f info-file] [-o output-file] [-n node-name]..."),
_("            [--directory dir-path] [--file info-file] [--node node-name]..."),
_("            [--help] [--output output-file] [--subnodes] [--version]"),
_("            [--dribble dribble-file] [--restore from-file]"),
_("            [menu-selection ...]"));
  exit (1);
}

/* Produce a scaled down description of the available options to Info. */
static void
info_short_help ()
{
  puts (_("\
Here is a quick description of Info's options.  For a more complete\n\
description of how to use Info, type `info info options'.\n\
\n\
   --directory DIR              Add DIR to INFOPATH.\n\
   --dribble FILENAME           Remember user keystrokes in FILENAME.\n\
   --file FILENAME              Specify Info file to visit.\n\
   --node NODENAME              Specify nodes in first visited Info file.\n\
   --output FILENAME            Output selected nodes to FILENAME.\n\
   --restore FILENAME           Read initial keystrokes from FILENAME.\n\
   --subnodes                   Recursively output menu items.\n\
   --help                       Get this help message.\n\
   --version                    Display Info's version information.\n\
\n\
Remaining arguments to Info are treated as the names of menu\n\
items in the initial node visited.  You can easily move to the\n\
node of your choice by specifying the menu names which describe\n\
the path to that node.  For example, `info emacs buffers'.\n\
\n\
Email bug reports to bug-texinfo@prep.ai.mit.edu."));

  exit (0);
}
